glusterd2/transactionv2/executor.go

```diff
@@ -0,0 +1,177 @@
+type executorImpl struct {
+	txnManager  TxnManager
+	stepManager StepManager
+	selfNodeID  uuid.UUID
+}
+
+// IsTxnPending returns true if the give txn is in pending state
+func (e *executorImpl) IsTxnPending(txnID uuid.UUID) bool {
+	status, err := e.txnManager.GetTxnStatus(txnID, e.selfNodeID)
+	if err != nil || status.State != txnPending {
+		return false
+	}
+	return true
+}
+
+// Execute will run all steps of a given txn on local Node. If a step is marked as synchornized,
+// then It will wait for all previous steps to complete on all involved Nodes.
+// If a node is an initiator node then It will acquire all cluster locks before running the txn steps.
+func (e *executorImpl) Execute(txn *Txn) error {
+	var (
+		errChan          = make(chan error)
+		done             = make(chan struct{})
+		updateStatusOnce = &sync.Once{}
+		ctx, cancel      = context.WithCancel(context.Background())
+	)
+	defer cancel()
+
+	if !e.shouldRunTxn(txn) {
+		return nil
+	}
+
+	if e.isInitiator(txn) {
+		if err := txn.acquireClusterLocks(); err != nil {
+			txn.Ctx.Logger().WithError(err).Error("error in acquiring cluster locks")
+			return err
+		}
+		defer txn.releaseLocks()
+	}
+
+	txn.Ctx.Logger().Info("transaction started on node")
+
+	failureChan := e.watchTxnForFailure(ctx, txn)
+
+	for i := range txn.Steps {
+		go e.runTxnStep(ctx, i, txn, errChan, done)
+
+		select {
+		case err := <-errChan:
+			status := TxnStatus{State: txnFailed, Reason: err.Error(), TxnID: txn.ID}
+			e.txnManager.UpDateTxnStatus(status, txn.ID, e.selfNodeID)
+			return err
+
+		case err := <-failureChan:
+			txn.Ctx.Logger().WithError(err).Error("transaction got failed on some other node, cancelling ongoing txn on this node")
+			return err
+
+		case <-done:
+			updateStatusOnce.Do(func() {
+				e.txnManager.UpDateTxnStatus(TxnStatus{State: txnRunning, TxnID: txn.ID}, txn.ID, e.selfNodeID)
+			})
+		}
+	}
+
+	return e.txnManager.UpDateTxnStatus(TxnStatus{State: txnSucceeded, TxnID: txn.ID}, txn.ID, e.selfNodeID)
+}
+
+// runTxnStep will run a txn Step having given `stepIndex`. On successful completion of step it will notify on given `done` chan.
+// In case of any error it will send notification of given `errChan`.
+// If a step is marked as sync then it will wait for all previous steps to complete on all involved nodes.
+func (e *executorImpl) runTxnStep(ctx context.Context, stepIndex int, txn *Txn, errChan chan<- error, done chan<- struct{}) {
+	var (
+		logger = txn.Ctx.Logger()
+		step   = txn.Steps[stepIndex]
+	)
+
+	logger.WithField("stepname", step.DoFunc).Debug("running step func on node")
+
+	// a synchronized step is executed only after all previous steps
+	// have been completed successfully by all involved peers.
+	if step.Sync {
+		logger.WithField("stepname", step.DoFunc).Debug("synchronizing txn step")
+		if err := e.stepManager.SyncStep(ctx, stepIndex, txn); err != nil {
+			logger.WithFields(log.Fields{
+				"error":    err,
```

Comment:
trivial: I think you can log this with  `.WithError(err)` just to maintain uniformity

Reply:
since i am using more than one field in log, so using WithError seems to be redundant, because WithError just add one field.

Reply:
I agree, but I think its a part of the convention that we follow. `logger.WithError(err).WithFields(log.Field("stepname": step.DoFunc)..Error("encounter an error in synchronizing txn step")`
should also work. This is not very important, you can skip it if you want.